backend/src/memberships/memberships.controller.ts

```diff
@@ -9,27 +9,121 @@ import {
+import { ChannelsService } from 'src/channels/channels.service';
+import { ChannelType } from 'src/channels/entities/channel.entity';
+import { EntityNotFoundError } from 'typeorm';
+import * as bcrypt from 'bcrypt';
+import { Request } from 'express';
+import { User } from 'src/users/entities/user.entity';
+
+async function userIsAdmin(
+  userId: number,
+  channelId: number,
+  membershipsService: MembershipsService,
+): Promise<boolean> {
+  const creatorMembership = await membershipsService.findAll({
+    user: userId.toString(),
+    channel: channelId.toString(),
+  });
+  if (
+    creatorMembership.length === 1 &&
+    (creatorMembership[0].role === MembershipRoleType.ADMIN ||
+      creatorMembership[0].role === MembershipRoleType.OWNER)
+  ) {
+    return true;
+  }
+  return false;
+}
+
+async function userIsOwner(
+  userId: number,
+  channelId: number,
+  membershipsService: MembershipsService,
+): Promise<boolean> {
+  const creatorMembership = await membershipsService.findAll({
+    user: userId.toString(),
+    channel: channelId.toString(),
+  });
+  if (
+    creatorMembership.length === 1 &&
+    creatorMembership[0].role === MembershipRoleType.OWNER
+  ) {
+    return true;
+  }
+  return false;
+}
 
 @ApiBearerAuth()
 @ApiTags('channel memberships')
 @Controller('memberships')
 export class MembershipsController {
-  constructor(private readonly membershipsService: MembershipsService) {}
+  constructor(
+    private readonly membershipsService: MembershipsService,
+    private readonly channelsService: ChannelsService,
+  ) {}
 
   @Post()
   async create(
     @Body() createMembershipDto: CreateMembershipDto,
+    @Req() req: Request,
   ): Promise<ResponseMembershipDto> {
+    try {
+      let isAuthorized = false;
+      const channel = await this.channelsService.findOne(
+        createMembershipDto.channelId,
+      );
+      if (createMembershipDto.role === MembershipRoleType.OWNER) {
+        throw new UnauthorizedException();
+      }
+      if (createMembershipDto.role === MembershipRoleType.ADMIN) {
+        if (
+          !(await userIsOwner(
+            (req.user as User).id,
+            channel.id,
+            this.membershipsService,
+          ))
+        ) {
+          throw new UnauthorizedException();
+        }
+      }
+      if (channel.type === ChannelType.PRIVATE) {
+        isAuthorized = await userIsAdmin(
+          (req.user as User).id,
+          createMembershipDto.channelId,
+          this.membershipsService,
+        );
+      } else if (channel.type === ChannelType.PROTECTED) {
+        isAuthorized =
+          createMembershipDto.password &&
+          (await bcrypt.compare(
+            createMembershipDto.password,
+            channel.password,
+          ));
+      } else {
+        isAuthorized = true;
+      }
+      if (!isAuthorized) {
+        throw new UnauthorizedException();
+      }
+    } catch (error) {
+      if (error instanceof EntityNotFoundError) {
```

Comment:
So this is related to line 85 findOne right ? But I think findOne returns undefined if nothing was found, and does not throw, so why not use findOneOrFail as you did in the service ?

Reply:
`channelsService.findOne` now uses `channelsRepository.findOneByOrFail` (since the TypeORM change), so it should throw now

Reply:
I often mistake the services findOne with the repositories'.. My bad